text/1015-full-text-search.rst

```diff
@@ -0,0 +1,372 @@
+``tsquery``.
+
+The ``tsvector`` type stores a processed version of the text. It basically
+contains information about the tokens making up the text well as their
+positions. This is the version of the text that is actually targeted by the
+FTS. Postgres also provides a few different ways to convert plain text into a
+``tsvector``. There's also a way to create an index that converts a text
+column into a ``tsvector`` and enables fast searches. Functions that convert
+text into ``tsvector`` can also take an argument that determines which parser
+will be used for the conversion.
+
+The ``tsquery`` type encapsulates the search conditions, such as which lexemes
+should or must appear and their relative order. Postgres also has several ways
+of converting a string into a ``tsquery``.
+
+Postgres provides a way to assign relative weights to different lexemes in the
+``tsvector`` which then can be used to rank FTS matches.
+
+
+Specification
+=============
+
+There's a lot of functionality involved in supporting FTS, but it's a fairly
+niche thing, so it makes sense to create a ``module fts`` for all of the FTS
+tools.
+
+In EdgeDB we should avoid having an explicit "document" scalar like
+``tsvector``. Instead we can rely on various parameters, annotations and
+indexes to specify how the plain text ``str`` should be parsed into tokens.
+This will allow us to keep the queries involving FTS mostly implementation
+agnostic. Some FTS backends may have functions specific to them, but the goal
+is to provide good coverage of common FTS functionality.
+
+In order to enable various way of manipulating the FTS query this proposal
+introduces ``ftsearch`` scalar type. It is meant to be analogous to
+``tsquery`` and it should be capable or representing common queries regardless
+of the specif FTS backend (Postgres, Elasticsearch, Lucene, etc.).
+
+The searching functionality can be split along two axis:
+1) what FTS operates on (``str`` vs ``Object``)
+2) what FTS returns (``bool`` vs some ranked results)
+
+The task of testing whether a document matches an FTS query would be performed
+by the following built-in functions::
+
+  function fts_test(doc: str, query: ftsearch) -> bool
+  function fts_test(doc: Object, query: ftsearch) -> bool
+
+The input can be either a ``str`` or an entire ``Object``. The object type
+must have some ``str`` properties, however. Trying to perform FTS on a type
+without ``str`` properties is an error. This function is generally intended to
+be used in ``filter`` clauses in cases where there's no need for hinting what
+matched, but the results are simple enough to be self-explanatory.
+
+The ranked results need to be rich objects with several relevant properties,
+akin to free objects used for returning ``group`` results. The difference is
+that we cannot use a ``FreeObject`` as a return type of a function, we need a
+specialized type for that: ``FTSResult``. This type is defined as follows::
+
+  type FTSResult {
```

Comment:
Not a huge fan of this composition.  Realistically, can we avoid including the input object/string in the result?  In practice the input would be a singleton (e.g. in a computed or in `filter`), so perhaps the result can be a named tuple instead: `<rank: float64, highlights: array<str>>`.

Example:

```
select Object
order by fts_match(Object).rank
limit 10
```

Reply:
Another point against a concrete type is that semantically those should be valid objects, i.e. they should exist and be queryable.

Reply:
Either one of the `rank` and especially `highlighs` (because highlighting is expensive and not always needed) can be empty. We need a placeholder for those cases, so we can use `0` for rank and `''` for no highlights.

Reply:
Sure, nothing wrong with having "zero" value here instead of an empty set

Reply:
Removed the copies of hits from `fts_match` results. Added defaults for `rank` and `highlights`.